### ISSUE TITLE
[KFNBC]: Unique and meaningful IDs for all elements in spawner/control panel

### ISSUE THREAD

**Issue by lugi0** [opendatahub-io/odh-dashboard#308]
### Feature description

To help with the integration of ods-ci, it would be great to have unique and meaningful IDs for all elements in both the spawner and control panel pages.
It would be even better if the IDs were the same as the ones currently used by the JH spawner page in RHODS.

**Comment by DaoDaoNoCode** [comment]
I will add `id` for all the elements once other features are implemented.

**Comment by lugi0** [comment]
Yes, DOM IDs. 
Using `data-id` is fine for me if that is the preferred field for React. No standard (at least that I know of) is being used right now, for us in QE the only request is that the IDs are unique and persistent so that we can use them to interact with the DOM. Even then IDs are not an absolute requirement, we are still able to somehow navigate around the DOM regardless, it just makes our lives easier (and our locators cleaner :).

If we could use the same field (`id`) with the same values we currently see in RHODS it would be a nice to have, simply because for us it might be easier to adapt our test framework as a drop-in replacement on KFNBC.

**Comment by andrewballantyne** [comment]
I highly recommend never using `id` properties for selecting elements in React. `id` is a unique property for the browser -- mid-page bookmarks use it for instance. 

I highly recommend we move to `data-` oriented properties as they have no expected use, so we can just agree on a convention. OpenShift Console uses `data-id`... but I don't have a problem with `data-test-id` if that's the preferred choice. We should however be as consistent as possible within' Open Data Hub Dashboard.

**Comment by andrewballantyne** [comment]
@lugi0 I can appreciate your position -- this is something we should have done earlier for sure. However, we have a lot of work to get done for this tight deadline early Aug. I'd like us to get ahead of the curve before we devote time to any refactorings, including adding test data attributes everywhere.

If you could provide a better idea of what doesn't have the test ids or what kinda tests you want to write, we could do a slim pass and give you those soon-ish. If your ask is "make it so we can select anything", it's important but unfortunately we need to start hitting deadlines with our features for the BU.

At the current rate, I'm thinking this is on the shelf for at least a week more.

cc @jkoehler-redhat -- Maybe you can chime in here about priorities and deadlines

**Comment by jkoehler-redhat** [comment]
@andrewballantyne @lugi0 I agree we have to decided what is absolutely necessary at this point.  We are not going to be able to do 100% coverage to meet this tight deadline.  @lugi0 can you please help to address what @andrewballantyne is mentioning for the slim pass?

**Comment by lugi0** [comment]
@andrewballantyne @jkoehler-redhat I understand there are important deadlines which will inevitably take priority over this, but the problem I am currently facing is that if I want to adapt our automation to work with KFNBC I run the risk of having to rework it multiple times until the UI/UX is finalized. 

To give a concrete example, the current UX for closing down a server is bound to change considerably from now to when KFNBC is shipped; if I want to run multiple suites of tests on this version of KFNBC, this procedure is fundamental to have, but I would have to spend effort to work around a UX which I already know will change. I am not sure the "wasted" effort is worth it in this case, but given the tight schedule we are all on it might become a necessity if we want automation in place when KFNBC is shipped to customers.

Another example would be the Log Out UX, which currently does not work; we have multiple tests that would become blocked by this; is the effort to work around this UX worth it vs. waiting for the UX to be changed?

Another example that could apply here is the spawn procedure itself. If I have understood correctly we are planning to overhaul the spawn UX to move to a modal rather than a full spawner page; if that were to happen, all work to adapt our tests from the JH spawner to the KFNBC spawner would likely be thrown out.

I'm hoping this paints an accurate picture of our problem; I think the closest would be @andrewballantyne 's previous comment, "make it so we can select anything", but not quite. While it is important for us to be able to select anything worth interacting with in the page(s), we are still able to navigate around the DOM if the IDs are not (yet) present, and we can always refactor our locators if IDs are added down the line; what concerns me the most is when will we be able to nail down the UX -  substantial changes in UX would require substantial rework in test automation as well, which could and likely will lead to automation not being ready on time.

I think we are unfortunately stuck in a chicken and egg situation here; I'd love to run our full automation ASAP to uncover any potential issue, but that comes with the baggage of having to work around the current UX while knowning the UX will surely change rather quickly, or at the very least by the time KFNBC is shipped, forcing us into a substantial rework.

If it can help in prioritizing, I could start adapting automation where it makes sense to do so (e.g. quick changes that don't require substantial reworks) in order to identify bigger blockers in our path, but I'm very much open to suggestions here
